src/main/scala/SchedulerImp.scala

```diff
@@ -0,0 +1,38 @@
+import java.util.concurrent.PriorityBlockingQueue
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SchedulerImp(
+    val minHeap: PriorityBlockingQueue[ScheduledTask]
+) extends Scheduler {
+
+  @volatile var running: Boolean = false
+
+  override def schedule(task: () => Unit, timestamp: Long): Unit = {
+    minHeap.offer(ScheduledTask(task, timestamp))
+  }
+
+  override def start()(implicit ec: ExecutionContext): Unit = {
+    if (!running) {
+      running = true
+      Future {
```

Comment:
The task-scheduler is now assigned its own thread, thus non-blocking the main thread.